prog/dftb+/lib_solvation/cosmo.F90

```diff
@@ -909,4 +913,131 @@ subroutine efld(nsrc, src, csrc, ntrg, ctrg, ef)
   end subroutine efld
 
 
+  !> Write a COSMO file output
+  subroutine writeCosmoFile(this, unit, species0, speciesNames, coords0, energy)
+
+    !> COSMO container
+    class(TCosmo), intent(in) :: this
+
+    !> Formatted unit for output
+    integer, intent(in) :: unit
+
+    !> Symbols of the species
+    character(len=*), intent(in) :: speciesNames(:)
+
+    !> Species of every atom in the unit cell
+    integer, intent(in) :: species0(:)
+
+    !> Atomic coordinates
+    real(dp), intent(in) :: coords0(:,:)
+
+    !> Total energy
+    real(dp), intent(in) :: energy
+
+    integer :: ii, ig, iat
+    real(dp) :: dielEnergy, keps
+    real(dp), allocatable :: phi(:), zeta(:), area(:)
+
+    allocate(phi(this%ddCosmo%ncav), zeta(this%ddCosmo%ncav), area(this%ddCosmo%ncav))
+    ! Reset potential on the cavity, note that the potential is expected in e/Å
+    call getPhi(this%chargesPerAtom, this%jmat, phi)
+    ii = 0
+    do iat = 1, this%ddCosmo%nat
+      do ig = 1, this%ddCosmo%ngrid
+        if (this%ddCosmo%ui(ig, iat) > 0.0_dp) then
+          ii = ii + 1
+          ! Calculate surface charge per area
+          zeta(ii) = this%ddCosmo%w(ig) * this%ddCosmo%ui(ig, iat) &
+              & * dot_product(this%ddCosmo%basis(:, ig), this%s(:, iat))
+          ! Save surface area in Ångström²
+          area(ii) = this%ddCosmo%w(ig) * Bohr__AA**2 * this%vdwRad(iat)**2
+        end if
+      end do
+    end do
+
+    ! Dielectric energy is the energy on the dielectric continuum
+    keps = 0.5_dp * (1.0_dp - 1.0_dp/this%dielectricConst)
+    dielEnergy = keps * dot_product(zeta, phi)
+
+    write(unit, '(a)') &
+        & "$info", &
+        & "prog.: dftb+"
+
+    write(unit, '(a)') &
+        & "$cosmo"
+    write(unit, '(2x, a:, "=", g0)') &
```

Comment:
Why is the colon edit descriptor needed here? The I/O list has a fixed size which matches the format descriptors, so no format rewinding happens.

Reply:
cosmo files can provide a string only entry here like `cavity closed`.